ostd/src/arch/x86/timer/apic.rs

```diff
@@ -30,14 +30,44 @@ use crate::{
 
 /// Initializes APIC with tsc deadline mode or periodic mode.
 /// Return the corresponding [`IrqLine`] for the System Timer.
-pub(super) fn init() -> IrqLine {
+pub(super) fn init_global_and_bsp() -> IrqLine {
     init_tsc_freq();
+
     if is_tsc_deadline_mode_supported() {
         info!("[Timer]: Enable APIC TSC deadline mode.");
-        init_tsc_mode()
+
+        let timer_irq = IrqLine::alloc().unwrap();
+
+        let callback = move || {
+            let tsc_value = unsafe { _rdtsc() };
+            let tsc_step = TSC_FREQ.load(Ordering::Relaxed) / TIMER_FREQ;
+            let next_tsc_value = tsc_step + tsc_value;
+            unsafe { wrmsr(IA32_TSC_DEADLINE, next_tsc_value) };
+        };
+
+        APIC_TIMER_CALLBACK.call_once(|| Arc::new(callback));
+
+        init_tsc_mode(&timer_irq);
+
+        timer_irq
     } else {
         info!("[Timer]: Enable APIC periodic mode.");
-        init_periodic_mode()
+
+        let mut timer_irq = IrqLine::alloc().unwrap();
+        timer_irq.on_active(pit_callback);
```

Comment:
Could you please test this branch? The previous code uses `pit_callback` (which is triggered by the PIT timer interrupt) to help calculate the frequency of the APIC timer interrupt. You cannot set `pit_callback` on the APIC timer handler, which is wrong and certainly will not work.